// probe.gl/dist/es6/lib/stat.js
import getHiResTimestamp from '../utils/hi-res-timestamp';
export default class Stat {
  constructor(name) {
    this.name = name;
    this.reset();
  }

  incrementCount() {
    this.addCount(1);
  }

  decrementCount() {
    this.subtractCount(1);
  }

  addCount(value) {
    this.count += value;
  }

  subtractCount(value) {
    this.count -= value;
  }

  addTime(time) {
    this.time += time;
    this.lastTiming = time;
    this.count++;
  }

  timeStart() {
    this._startTime = getHiResTimestamp();
    this._timerPending = true;
  }

  timeEnd() {
    if (!this._timerPending) {
      return;
    }

    this.addTime(getHiResTimestamp() - this._startTime);
    this._timerPending = false;
  }

  getAverageTime() {
    return this.count > 0 ? this.time / this.count : 0;
  }

  getHz() {
    return this.time > 0 ? this.count / (this.time / 1000) : 0;
  }

  reset() {
    this.time = 0;
    this.count = 0;
    this.lastTiming = 0;
    this._startTime = 0;
    this._timerPending = false;
  }

}
