// @loaders.gl/core/dist/es6/lib/parse.js
import _asyncToGenerator from "@babel/runtime/helpers/esm/asyncToGenerator";
import { autoDetectLoader } from './loader-utils/auto-detect-loader';
import { normalizeLoader, isLoaderObject } from './loader-utils/normalize-loader';
import NullLog from './loader-utils/null-log';
import { getRegisteredLoaders } from './register-loaders';
import { parseWithLoader, parseWithLoaderInBatches, parseWithLoaderSync } from './parse-with-loader';
export function parse(_x, _x2, _x3, _x4) {
  return _parse.apply(this, arguments);
}

function _parse() {
  _parse = _asyncToGenerator(function* (data, loaders, options, url) {
    if (!Array.isArray(loaders) && !isLoaderObject(loaders)) {
      url = options;
      options = loaders;
      loaders = null;
    }

    loaders = loaders || getRegisteredLoaders();
    const loader = Array.isArray(loaders) ? autoDetectLoader(url, data, loaders) : loaders;

    if (!loader) {
      return null;
    }

    normalizeLoader(loader);
    options = mergeLoaderAndUserOptions(options, loader);
    return yield parseWithLoader(data, loader, options, url);
  });
  return _parse.apply(this, arguments);
}

export function parseSync(data, loaders, options, url) {
  if (!Array.isArray(loaders) && !isLoaderObject(loaders)) {
    url = options;
    options = loaders;
    loaders = null;
  }

  loaders = loaders || getRegisteredLoaders();
  const loader = Array.isArray(loaders) ? autoDetectLoader(url, data, loaders) : loaders;
  normalizeLoader(loader);
  options = mergeLoaderAndUserOptions(options, loader);
  return parseWithLoaderSync(data, loader, options, url);
}
export function parseInBatches(_x5, _x6, _x7, _x8) {
  return _parseInBatches.apply(this, arguments);
}

function _parseInBatches() {
  _parseInBatches = _asyncToGenerator(function* (data, loaders, options, url) {
    if (!Array.isArray(loaders) && !isLoaderObject(loaders)) {
      url = options;
      options = loaders;
      loaders = null;
    }

    loaders = loaders || getRegisteredLoaders();
    const loader = Array.isArray(loaders) ? autoDetectLoader(url, null, loaders) : loaders;
    normalizeLoader(loader);
    options = mergeLoaderAndUserOptions(options, loader);
    return parseWithLoaderInBatches(data, loader, options, url);
  });
  return _parseInBatches.apply(this, arguments);
}

export function parseInBatchesSync(_x9, _x10, _x11, _x12) {
  return _parseInBatchesSync.apply(this, arguments);
}

function _parseInBatchesSync() {
  _parseInBatchesSync = _asyncToGenerator(function* (data, loaders, options, url) {
    if (!Array.isArray(loaders) && !isLoaderObject(loaders)) {
      url = options;
      options = loaders;
      loaders = null;
    }

    loaders = loaders || getRegisteredLoaders();
    const loader = Array.isArray(loaders) ? autoDetectLoader(url, null, loaders) : loaders;
    normalizeLoader(loader);
    options = mergeLoaderAndUserOptions(options, loader);
    return parseWithLoaderInBatches(data, loader, options, url);
  });
  return _parseInBatchesSync.apply(this, arguments);
}

function mergeLoaderAndUserOptions(options, loader) {
  options = Object.assign({}, loader.DEFAULT_OPTIONS, loader.options, options, {
    dataType: 'arraybuffer'
  });

  if (options.log === null) {
    options.log = new NullLog();
  }

  if (!('log' in options)) {
    options.log = console;
  }

  return options;
}
