// @deck.gl/mapbox/dist/esm/mapbox-layer.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import { getDeckInstance, addLayer, removeLayer, updateLayer, drawLayer, getViewState } from './deck-utils';

var MapboxLayer = function () {
  function MapboxLayer(props) {
    _classCallCheck(this, MapboxLayer);

    if (!props.id) {
      throw new Error('Layer must have an unique id');
    }

    this.id = props.id;
    this.type = 'custom';
    this.renderingMode = props.renderingMode || '3d';
    this.map = null;
    this.deck = null;
    this.props = props;
  }

  _createClass(MapboxLayer, [{
    key: "onAdd",
    value: function onAdd(map, gl) {
      this.map = map;
      this.deck = getDeckInstance({
        map: map,
        gl: gl,
        deck: this.props.deck
      });
      addLayer(this.deck, this);
    }
  }, {
    key: "onRemove",
    value: function onRemove() {
      removeLayer(this.deck, this);
    }
  }, {
    key: "setProps",
    value: function setProps(props) {
      Object.assign(this.props, props, {
        id: this.id
      });

      if (this.deck) {
        updateLayer(this.deck, this);
      }
    }
  }, {
    key: "render",
    value: function render(gl, matrix) {
      this.deck.setProps({
        viewState: this._getViewState()
      });
      drawLayer(this.deck, this);
    }
  }, {
    key: "_getViewState",
    value: function _getViewState() {
      var map = this.map,
          deck = this.deck;
      return getViewState(map, {
        nearZMultiplier: deck.height ? 1 / deck.height : 1,
        farZMultiplier: 1
      });
    }
  }]);

  return MapboxLayer;
}();

export { MapboxLayer as default };
