// @deck.gl/aggregation-layers/dist/es5/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ScreenGridLayer", {
  enumerable: true,
  get: function get() {
    return _screenGridLayer.default;
  }
});
Object.defineProperty(exports, "GridLayer", {
  enumerable: true,
  get: function get() {
    return _gridLayer.default;
  }
});
Object.defineProperty(exports, "HexagonLayer", {
  enumerable: true,
  get: function get() {
    return _hexagonLayer.default;
  }
});
Object.defineProperty(exports, "ContourLayer", {
  enumerable: true,
  get: function get() {
    return _contourLayer.default;
  }
});
Object.defineProperty(exports, "_GPUGridLayer", {
  enumerable: true,
  get: function get() {
    return _gpuGridLayer.default;
  }
});
Object.defineProperty(exports, "_GPUGridAggregator", {
  enumerable: true,
  get: function get() {
    return _gpuGridAggregator.default;
  }
});
Object.defineProperty(exports, "AGGREGATION_OPERATION", {
  enumerable: true,
  get: function get() {
    return _gpuGridAggregatorConstants.AGGREGATION_OPERATION;
  }
});
Object.defineProperty(exports, "_pointToDensityGridData", {
  enumerable: true,
  get: function get() {
    return _gridAggregationUtils.pointToDensityGridData;
  }
});
exports.experimental = void 0;

var _screenGridLayer = _interopRequireDefault(require("./screen-grid-layer/screen-grid-layer"));

var _gridLayer = _interopRequireDefault(require("./grid-layer/grid-layer"));

var _hexagonLayer = _interopRequireDefault(require("./hexagon-layer/hexagon-layer"));

var _contourLayer = _interopRequireDefault(require("./contour-layer/contour-layer"));

var _gpuGridLayer = _interopRequireDefault(require("./gpu-grid-layer/gpu-grid-layer"));

var _gpuGridAggregator = _interopRequireDefault(require("./utils/gpu-grid-aggregation/gpu-grid-aggregator"));

var _gpuGridAggregatorConstants = require("./utils/gpu-grid-aggregation/gpu-grid-aggregator-constants");

var _gridAggregationUtils = require("./utils/gpu-grid-aggregation/grid-aggregation-utils");

var _binSorter = _interopRequireDefault(require("./utils/bin-sorter"));

var _scaleUtils = require("./utils/scale-utils");

var _colorUtils = require("./utils/color-utils");

var experimental = {
  BinSorter: _binSorter.default,
  linearScale: _scaleUtils.linearScale,
  getLinearScale: _scaleUtils.getLinearScale,
  quantizeScale: _scaleUtils.quantizeScale,
  getQuantizeScale: _scaleUtils.getQuantizeScale,
  defaultColorRange: _colorUtils.defaultColorRange
};
exports.experimental = experimental;
