// @loaders.gl/images/dist/es6/lib/get-image-metadata.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { bufferToArrayBuffer } from '../node/buffer-to-array-buffer';
const BIG_ENDIAN = false;
const LITTLE_ENDIAN = true;
const mimeTypeMap = new Map([['image/png', getPngSize], ['image/jpeg', getJpegSize], ['image/gif', getGifSize], ['image/bmp', getBmpSize]]);
const ERR_INVALID_TYPE = "Invalid MIME type. Supported MIME types are: ".concat(Array.from(mimeTypeMap.keys()).join(', '));
export function isImage(arrayBuffer) {
  const result = guessImageMetadata(arrayBuffer);
  return result ? result.mimeType : false;
}
export function getImageMetadata(arrayBuffer) {
  let mimeType = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;

  if (mimeType) {
    const handler = mimeTypeMap.get(mimeType);

    if (!handler) {
      throw new Error(ERR_INVALID_TYPE);
    }

    const result = handler(arrayBuffer);

    if (!result) {
      throw new Error("invalid image data for type: ".concat(mimeType));
    }

    return result;
  }

  const result = guessImageMetadata(arrayBuffer, mimeType);

  if (!result) {
    throw new Error(ERR_INVALID_TYPE);
  }

  return result;
}

function guessImageMetadata(arrayBuffer, mimeType) {
  for (const _ref of mimeTypeMap.entries()) {
    var _ref2 = _slicedToArray(_ref, 2);

    const supportedMimeType = _ref2[0];
    const handler = _ref2[1];
    const result = handler(arrayBuffer);

    if (result) {
      result.mimeType = supportedMimeType;
      return result;
    }
  }

  return null;
}

function getPngSize(arrayBuffer) {
  const dataView = toDataView(arrayBuffer);

  if (dataView.byteLength < 24 || dataView.getUint32(0, BIG_ENDIAN) !== 0x89504e47) {
    return null;
  }

  return {
    width: dataView.getUint32(16, BIG_ENDIAN),
    height: dataView.getUint32(20, BIG_ENDIAN)
  };
}

function getGifSize(arrayBuffer) {
  const dataView = toDataView(arrayBuffer);

  if (dataView.byteLength < 10 || dataView.getUint32(0, BIG_ENDIAN) !== 0x47494638) {
    return null;
  }

  return {
    width: dataView.getUint16(6, LITTLE_ENDIAN),
    height: dataView.getUint16(8, LITTLE_ENDIAN)
  };
}

function getBmpSize(arrayBuffer) {
  const dataView = toDataView(arrayBuffer);

  if (dataView.getUint16(0, BIG_ENDIAN) !== 0x424d) {
    return null;
  }

  return {
    width: dataView.getUint32(18, LITTLE_ENDIAN),
    height: dataView.getUint32(22, LITTLE_ENDIAN)
  };
}

function getJpegSize(arrayBuffer) {
  const dataView = toDataView(arrayBuffer);

  if (dataView.byteLength < 2 || dataView.getUint16(0, BIG_ENDIAN) !== 0xffd8) {
    return null;
  }

  const _getJpegMarkers = getJpegMarkers(),
        tableMarkers = _getJpegMarkers.tableMarkers,
        sofMarkers = _getJpegMarkers.sofMarkers;

  let i = 2;

  while (i < dataView.byteLength) {
    const marker = dataView.getUint16(i, BIG_ENDIAN);

    if (sofMarkers.has(marker)) {
      return {
        height: dataView.getUint16(i + 5, BIG_ENDIAN),
        width: dataView.getUint16(i + 7, BIG_ENDIAN)
      };
    }

    if (!tableMarkers.has(marker)) {
      return null;
    }

    i += 2;
    i += dataView.getUint16(i, BIG_ENDIAN);
  }

  return null;
}

function getJpegMarkers() {
  const tableMarkers = new Set([0xffdb, 0xffc4, 0xffcc, 0xffdd, 0xfffe]);

  for (let i = 0xffe0; i < 0xfff0; ++i) {
    tableMarkers.add(i);
  }

  const sofMarkers = new Set([0xffc0, 0xffc1, 0xffc2, 0xffc3, 0xffc5, 0xffc6, 0xffc7, 0xffc9, 0xffca, 0xffcb, 0xffcd, 0xffce, 0xffcf, 0xffde]);
  return {
    tableMarkers,
    sofMarkers
  };
}

function toDataView(data) {
  if (bufferToArrayBuffer) {
    data = bufferToArrayBuffer(data);
  }

  if (data instanceof ArrayBuffer || ArrayBuffer.isView(data)) {
    return new DataView(data.buffer || data);
  }

  throw new Error('toDataView');
}

export function getImageSize(arrayBuffer) {
  let mimeType = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
  return getImageMetadata(arrayBuffer);
}
