// @loaders.gl/core/dist/es5/lib/loader-utils/get-data.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getUrlFromData = getUrlFromData;
exports.getSizeFromData = getSizeFromData;
exports.getArrayBufferOrStringFromDataSync = getArrayBufferOrStringFromDataSync;
exports.getArrayBufferOrStringFromData = getArrayBufferOrStringFromData;
exports.getAsyncIteratorFromData = getAsyncIteratorFromData;
exports.getIteratorFromData = getIteratorFromData;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _isType = require("../../javascript-utils/is-type");

var _streamUtils = require("../../javascript-utils/stream-utils");

var _textEncoding = require("../../javascript-utils/text-encoding");

var ERR_DATA = 'Cannot convert supplied data type';

function getUrlFromData(data) {
  return (0, _isType.isFetchResponse)(data) ? data.url : null;
}

function getSizeFromData(data) {
  return (0, _isType.isFetchResponse)(data) ? data.headers.get('Content-Length') : null;
}

function getArrayBufferOrStringFromDataSync(data, loader) {
  if (loader.text && typeof data === 'string') {
    return data;
  }

  if (data instanceof ArrayBuffer || ArrayBuffer.isView(data)) {
    var arrayBuffer = data.buffer || data;

    if (loader.text && !loader.binary) {
      var textDecoder = new _textEncoding.TextDecoder('utf8');
      return textDecoder.decode(arrayBuffer);
    }

    return arrayBuffer;
  }

  throw new Error(ERR_DATA);
}

function getArrayBufferOrStringFromData(_x, _x2) {
  return _getArrayBufferOrStringFromData.apply(this, arguments);
}

function _getArrayBufferOrStringFromData() {
  _getArrayBufferOrStringFromData = (0, _asyncToGenerator2["default"])(_regenerator["default"].mark(function _callee(data, loader) {
    var isArrayBuffer;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return data;

          case 2:
            data = _context.sent;
            isArrayBuffer = data instanceof ArrayBuffer || ArrayBuffer.isView(data);

            if (!(typeof data === 'string' || isArrayBuffer)) {
              _context.next = 6;
              break;
            }

            return _context.abrupt("return", getArrayBufferOrStringFromDataSync(data, loader));

          case 6:
            if (!(0, _isType.isFetchResponse)(data)) {
              _context.next = 15;
              break;
            }

            if (!loader.binary) {
              _context.next = 13;
              break;
            }

            _context.next = 10;
            return data.arrayBuffer();

          case 10:
            _context.t0 = _context.sent;
            _context.next = 14;
            break;

          case 13:
            _context.t0 = data.text();

          case 14:
            return _context.abrupt("return", _context.t0);

          case 15:
            throw new Error(ERR_DATA);

          case 16:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _getArrayBufferOrStringFromData.apply(this, arguments);
}

function getAsyncIteratorFromData(_x3, _x4) {
  return _getAsyncIteratorFromData.apply(this, arguments);
}

function _getAsyncIteratorFromData() {
  _getAsyncIteratorFromData = (0, _asyncToGenerator2["default"])(_regenerator["default"].mark(function _callee2(data, loader) {
    return _regenerator["default"].wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            if (!(0, _isType.isIterator)(data)) {
              _context2.next = 2;
              break;
            }

            return _context2.abrupt("return", data);

          case 2:
            if (!(0, _isType.isFetchResponse)(data)) {
              _context2.next = 4;
              break;
            }

            return _context2.abrupt("return", (0, _streamUtils.getStreamIterator)(data.body));

          case 4:
            if (!(0, _isType.isReadableStream)(data)) {
              _context2.next = 6;
              break;
            }

            return _context2.abrupt("return", (0, _streamUtils.getStreamIterator)(data));

          case 6:
            if (!(0, _isType.isAsyncIterable)(data)) {
              _context2.next = 8;
              break;
            }

            return _context2.abrupt("return", data[Symbol.asyncIterator]());

          case 8:
            return _context2.abrupt("return", getIteratorFromData(data, loader));

          case 9:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));
  return _getAsyncIteratorFromData.apply(this, arguments);
}

function getIteratorFromData(_x5, _x6) {
  return _getIteratorFromData.apply(this, arguments);
}

function _getIteratorFromData() {
  _getIteratorFromData = (0, _asyncToGenerator2["default"])(_regenerator["default"].mark(function _callee3(data, loader) {
    return _regenerator["default"].wrap(function _callee3$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            if (!(data instanceof ArrayBuffer || ArrayBuffer.isView(data))) {
              _context4.next = 2;
              break;
            }

            return _context4.abrupt("return", _regenerator["default"].mark(function oneChunk() {
              return _regenerator["default"].wrap(function oneChunk$(_context3) {
                while (1) {
                  switch (_context3.prev = _context3.next) {
                    case 0:
                      _context3.next = 2;
                      return data.buffer || data;

                    case 2:
                    case "end":
                      return _context3.stop();
                  }
                }
              }, oneChunk);
            })());

          case 2:
            if (!(0, _isType.isIterator)(data)) {
              _context4.next = 4;
              break;
            }

            return _context4.abrupt("return", data);

          case 4:
            if (!(0, _isType.isIterable)(data)) {
              _context4.next = 6;
              break;
            }

            return _context4.abrupt("return", data[Symbol.iterator]());

          case 6:
            throw new Error(ERR_DATA);

          case 7:
          case "end":
            return _context4.stop();
        }
      }
    }, _callee3);
  }));
  return _getIteratorFromData.apply(this, arguments);
}
