// @loaders.gl/core/dist/esm/node/fetch/fetch-file-node.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread";
import _regeneratorRuntime from "@babel/runtime/regenerator";
import _asyncToGenerator from "@babel/runtime/helpers/esm/asyncToGenerator";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import fs from 'fs';
import http from 'http';
import https from 'https';
import util from 'util';
import { toArrayBuffer } from '../../javascript-utils/binary-utils';
import { TextDecoder } from '../../javascript-utils/text-encoding';
import { concatenateReadStream } from '../../javascript-utils/stream-utils';
import decodeDataUri from './decode-data-uri';
var DEFAULT_OPTIONS = {
  dataType: 'arrayBuffer'
};
var isNode = Boolean(fs && fs.readFile);

var isDataURL = function isDataURL(url) {
  return url.startsWith('data:');
};

var isRequestURL = function isRequestURL(url) {
  return url.startsWith('http:') || url.startsWith('https:');
};

var NodeHeaders = function () {
  function NodeHeaders(response) {
    _classCallCheck(this, NodeHeaders);

    this.reponse = response;
  }

  _createClass(NodeHeaders, [{
    key: "get",
    value: function get(header) {
      if (this.response.httpResponse) {
        return this.response.httpResponse.getHeaders()[header];
      }

      switch (header) {
        case 'Content-Length':
          return this._getContentLength();

        default:
      }

      return undefined;
    }
  }, {
    key: "_getContentLength",
    value: function _getContentLength() {
      return this._contentLength;
    }
  }]);

  return NodeHeaders;
}();

var NodeFetchResponse = function () {
  function NodeFetchResponse(url, options) {
    _classCallCheck(this, NodeFetchResponse);

    this.url = url;
    this.options = options;
    this.bodyUsed = false;
    this._headers = null;
  }

  _createClass(NodeFetchResponse, [{
    key: "arrayBuffer",
    value: function () {
      var _arrayBuffer = _asyncToGenerator(_regeneratorRuntime.mark(function _callee() {
        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                this.bodyUsed = true;
                return _context.abrupt("return", readFile(this.url, this.options));

              case 2:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function arrayBuffer() {
        return _arrayBuffer.apply(this, arguments);
      }

      return arrayBuffer;
    }()
  }, {
    key: "text",
    value: function () {
      var _text = _asyncToGenerator(_regeneratorRuntime.mark(function _callee2() {
        var arrayBuffer, textDecoder;
        return _regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this.arrayBuffer();

              case 2:
                arrayBuffer = _context2.sent;
                textDecoder = new TextDecoder();
                return _context2.abrupt("return", textDecoder.decode(arrayBuffer));

              case 5:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function text() {
        return _text.apply(this, arguments);
      }

      return text;
    }()
  }, {
    key: "json",
    value: function () {
      var _json = _asyncToGenerator(_regeneratorRuntime.mark(function _callee3() {
        var text;
        return _regeneratorRuntime.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return this.text();

              case 2:
                text = _context3.sent;
                return _context3.abrupt("return", JSON.parse(text));

              case 4:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function json() {
        return _json.apply(this, arguments);
      }

      return json;
    }()
  }, {
    key: "_getHeaders",
    value: function _getHeaders() {
      if (isRequestURL(this.url)) {
        return new NodeHeaders(this);
      }

      if (isDataURL(this.url)) {
        return {
          'Content-Length': this.url.length
        };
      }

      if (isRequestURL(this.repsonse.url)) {
        return {};
      }

      var stats = fs.statSync(this.response.url);
      return {
        'Content-Length': stats.size
      };
    }
  }, {
    key: "headers",
    get: function get() {
      this._headers = this._headers || this._getHeaders();
    }
  }, {
    key: "body",
    get: function get() {
      var url = this.url,
          options = this.options;
      this.bodyUsed = true;
      return createReadStream(url, options);
    }
  }]);

  return NodeFetchResponse;
}();

export function fetchFile(_x, _x2) {
  return _fetchFile.apply(this, arguments);
}

function _fetchFile() {
  _fetchFile = _asyncToGenerator(_regeneratorRuntime.mark(function _callee4(url, options) {
    return _regeneratorRuntime.wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            return _context4.abrupt("return", new NodeFetchResponse(url, options));

          case 1:
          case "end":
            return _context4.stop();
        }
      }
    }, _callee4);
  }));
  return _fetchFile.apply(this, arguments);
}

export function readFileSync(url) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  options = getReadFileOptions(options);

  if (isDataURL(url)) {
    return decodeDataUri(url);
  }

  if (!isNode) {
    return null;
  }

  var buffer = fs.readFileSync(url, options, function () {});
  return buffer instanceof Buffer ? toArrayBuffer(buffer) : buffer;
}

function readFile(_x3) {
  return _readFile.apply(this, arguments);
}

function _readFile() {
  _readFile = _asyncToGenerator(_regeneratorRuntime.mark(function _callee5(url) {
    var options,
        readFileAsync,
        buffer,
        _args5 = arguments;
    return _regeneratorRuntime.wrap(function _callee5$(_context5) {
      while (1) {
        switch (_context5.prev = _context5.next) {
          case 0:
            options = _args5.length > 1 && _args5[1] !== undefined ? _args5[1] : {};
            options = getReadFileOptions(options);

            if (!isDataURL(url)) {
              _context5.next = 4;
              break;
            }

            return _context5.abrupt("return", Promise.resolve(decodeDataUri(url)));

          case 4:
            if (!isRequestURL(url)) {
              _context5.next = 6;
              break;
            }

            return _context5.abrupt("return", new Promise(function (resolve, reject) {
              options = _objectSpread({}, new URL(url), options);
              var request = url.startsWith('https:') ? https.request : http.request;
              request(url, function (response) {
                return concatenateReadStream(response).then(resolve, reject);
              });
            }));

          case 6:
            readFileAsync = util.promisify(fs.readFile);
            _context5.next = 9;
            return readFileAsync(url, options);

          case 9:
            buffer = _context5.sent;
            return _context5.abrupt("return", buffer instanceof Buffer ? toArrayBuffer(buffer) : buffer);

          case 11:
          case "end":
            return _context5.stop();
        }
      }
    }, _callee5);
  }));
  return _readFile.apply(this, arguments);
}

export function createReadStream(_x4, _x5) {
  return _createReadStream.apply(this, arguments);
}

function _createReadStream() {
  _createReadStream = _asyncToGenerator(_regeneratorRuntime.mark(function _callee6(url, options) {
    return _regeneratorRuntime.wrap(function _callee6$(_context6) {
      while (1) {
        switch (_context6.prev = _context6.next) {
          case 0:
            if (!isDataURL(url)) {
              _context6.next = 2;
              break;
            }

            return _context6.abrupt("return", decodeDataUri(url));

          case 2:
            if (isRequestURL(url)) {
              _context6.next = 4;
              break;
            }

            return _context6.abrupt("return", fs.createReadStream(url, options));

          case 4:
            return _context6.abrupt("return", new Promise(function (resolve, reject) {
              options = _objectSpread({}, new URL(url), options);
              var request = url.startsWith('https:') ? https.request : http.request;
              request(url, function (response) {
                return resolve(response);
              });
            }));

          case 5:
          case "end":
            return _context6.stop();
        }
      }
    }, _callee6);
  }));
  return _createReadStream.apply(this, arguments);
}

function getReadFileOptions() {
  var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  options = Object.assign({}, DEFAULT_OPTIONS, options);
  options.responseType = options.responseType || options.dataType;

  if (fs) {
    options.encoding = options.encoding || (options.dataType === 'text' ? 'utf8' : null);
  }

  return options;
}
